### ISSUE TITLE
Training Details

### ISSUE THREAD

**Issue by rongakowang** [IceClear/StableSR#22]
Hi,

Thank you for providing the training code. As I am trying to retrain the model, I wish to clarify a few details below:

1) When downloading the `v2-1_512-ema-pruned.ckpt` (from [here](https://huggingface.co/stabilityai/stable-diffusion-2-1-base/resolve/main/v2-1_512-ema-pruned.ckpt)) as instructed, there are some keys unexpected, I wonder what are they used for?

```
Unexpected Keys: ['model_ema.decay', 'model_ema.num_updates']
``` 

2) In the implementation the SFT layers is replaced with SPADE layers, I wonder what is the difference between them?

3) I notice that multiple times the model is attempting to check `hasattr(self, "split_input_params")`, which is so far as I observed always False, may I know what is this argument used for? When it is True, it seems to use `torch.nn.Fold` to encode rather than the VAE encoder.

4) It seems in training the paper uses DIV2K with HR resolution 512x512. However, the initial data has resolution 2k, I wonder what (may be resize) is done to fit this configuration? 

5) Do we train in aggregation sampling? it seems in Appendix B it does not have learnable parameters.

6) May I know what are the training losses? In particular do you have image based loss (on pixels, rather than the latents)? If so, how do you handle the case mentioned in 4). with training image resolution > 512?

Thank you in advance for your time and answering!

**Comment by IceClear** [comment]
Hi. Thanks for your interest of our work.
1. They are for ema training, used by SD.
2. In my opinion, no difference. I think they have the same math formulation. I may change the implementation to avoid misunderstanding.
3. This is used by SD, not StableSR. I just keep it as the original SD.
4. Crop, you may refer to RealESRGAN to check common settings of SR.
5. It is just for sampling.
6. The same as SD.

**Comment by rongakowang** [comment]
Hi @IceClear:

Thank you for your answer! 

For some further questions:
4. So in this case during testing, for image with resolution > 512, we are just cropping images of size 512x512 with overlapping and use the aggregation sampling to obtain the output with the desired resolution, which refers to the `tiling` strategy?
6. In the second stage, when you train the CFW what are the training losses? Since the entire diffusion model in the previous stage is freezed, I am assuming other losses are used here?

**Comment by IceClear** [comment]
4. Yes.
5. 
![image](https://github.com/IceClear/StableSR/assets/22350795/87e3ea6f-9158-4b91-94e9-42ca11c4f5e4)
You may refer to the code for details.

**Comment by rongakowang** [comment]
Hi @IceClear:

Thank you for your answer, I have tried to train the model on the DIV2K training set, but found the loss curve as below:

<img width="1200" alt="屏幕快照 2023-05-30 12 05 26" src="https://github.com/IceClear/StableSR/assets/45116569/6789cf1d-ce45-416e-8a26-f84f7bf46c76">

with the training config:

```
sf: 4
model:
  base_learning_rate: 5.0e-05
  target: ldm.models.diffusion.ddpm.LatentDiffusionSRTextWT
  params:
    # parameterization: "v"
    linear_start: 0.00085
    linear_end: 0.0120
    num_timesteps_cond: 1
    log_every_t: 200
    timesteps: 1000
    first_stage_key: image
    cond_stage_key: caption
    image_size: 512
    channels: 4
    cond_stage_trainable: False   # Note: different from the one we trained before
    conditioning_key: crossattn
    monitor: val/loss_simple_ema
    scale_factor: 0.18215
    use_ema: False
    # for training only
    # ckpt_path: /mnt/lustre/jywang/code/stable_diffmodels/v2-1_512-ema-pruned.ckpt
    ckpt_path: /home/rongakowang/StableSR/ckpts/v2-1_512-ema-pruned.ckpt
    unfrozen_diff: False
    random_size: False
    time_replace: 1000
    use_usm: True
    #P2 weighting, we do not use in final version
    p2_gamma: ~
    p2_k: ~
    # ignore_keys: []

    unet_config:
      target: ldm.modules.diffusionmodules.openaimodel.UNetModelDualcondV2
      params:
        image_size: 32 # unused
        in_channels: 4
        out_channels: 4
        model_channels: 320
        attention_resolutions: [ 4, 2, 1 ]
        num_res_blocks: 2
        channel_mult: [ 1, 2, 4, 4 ]
        num_head_channels: 64
        use_spatial_transformer: True
        use_linear_in_transformer: True
        transformer_depth: 1
        context_dim: 1024
        use_checkpoint: False
        legacy: False
        semb_channels: 256

    first_stage_config:
      target: ldm.models.autoencoder.AutoencoderKL
      params:
        # for training only
        # ckpt_path: /mnt/lustre/jywang/code/stable_diffmodels/v2-1_512-ema-pruned.ckpt
        ckpt_path: /home/rongakowang/StableSR/ckpts/v2-1_512-ema-pruned.ckpt
        embed_dim: 4
        monitor: val/rec_loss
        ddconfig:
          double_z: true
          z_channels: 4
          resolution: 512
          in_channels: 3
          out_ch: 3
          ch: 128
          ch_mult:
          - 1
          - 2
          - 4
          - 4
          num_res_blocks: 2
          attn_resolutions: []
          dropout: 0.0
        lossconfig:
          target: torch.nn.Identity

    cond_stage_config:
      target: ldm.modules.encoders.modules.FrozenOpenCLIPEmbedder
      params:
        freeze: True
        layer: "penultimate"

    structcond_stage_config:
      target: ldm.modules.diffusionmodules.openaimodel.EncoderUNetModelWT
      params:
        image_size: 96
        in_channels: 4
        model_channels: 256
        out_channels: 256
        num_res_blocks: 2
        attention_resolutions: [ 4, 2, 1 ]
        dropout: 0
        channel_mult: [ 1, 1, 2, 2 ]
        conv_resample: True
        dims: 2
        use_checkpoint: False
        use_fp16: False
        num_heads: 4
        num_head_channels: -1
        num_heads_upsample: -1
        use_scale_shift_norm: False
        resblock_updown: False
        use_new_attention_order: False


degradation:
  # the first degradation process
  resize_prob: [0.2, 0.7, 0.1]  # up, down, keep
  resize_range: [0.3, 1.5]
  gaussian_noise_prob: 0.5
  noise_range: [1, 15]
  poisson_scale_range: [0.05, 2.0]
  gray_noise_prob: 0.4
  jpeg_range: [60, 95]

  # the second degradation process
  second_blur_prob: 0.5
  resize_prob2: [0.3, 0.4, 0.3]  # up, down, keep
  resize_range2: [0.6, 1.2]
  gaussian_noise_prob2: 0.5
  noise_range2: [1, 12]
  poisson_scale_range2: [0.05, 1.0]
  gray_noise_prob2: 0.4
  jpeg_range2: [60, 100]

  gt_size: 512
  no_degradation_prob: 0.01

data:
  target: main.DataModuleFromConfig
  params:
    batch_size: 2
    num_workers: 6
    wrap: false
    train:
      target: basicsr.data.realesrgan_dataset.RealESRGANDataset
      params:
        queue_size: 180
        gt_path: ['/home/rongakowang/StableSR/local_data/DIV2K_train_HR']
        # face_gt_path: '/mnt/lustre/share/jywang/dataset/FFHQ/1024/'
        # num_face: 10000
        crop_size: 512
        io_backend:
          type: disk

        blur_kernel_size: 21
        kernel_list: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob: 0.1
        blur_sigma: [0.2, 1.5]
        betag_range: [0.5, 2.0]
        betap_range: [1, 1.5]

        blur_kernel_size2: 11
        kernel_list2: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob2: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob2: 0.1
        blur_sigma2: [0.2, 1.0]
        betag_range2: [0.5, 2.0]
        betap_range2: [1, 1.5]

        final_sinc_prob: 0.8

        gt_size: 512
        use_hflip: True
        use_rot: False
    validation:
      target: basicsr.data.realesrgan_dataset.RealESRGANDataset
      params:
        gt_path: /home/rongakowang/StableSR/local_data/DIV2K_valid_HR
        crop_size: 512
        io_backend:
          type: disk

        blur_kernel_size: 21
        kernel_list: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob: 0.1
        blur_sigma: [0.2, 1.5]
        betag_range: [0.5, 2.0]
        betap_range: [1, 1.5]

        blur_kernel_size2: 11
        kernel_list2: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob2: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob2: 0.1
        blur_sigma2: [0.2, 1.0]
        betag_range2: [0.5, 2.0]
        betap_range2: [1, 1.5]

        final_sinc_prob: 0.8

        gt_size: 512
        use_hflip: True
        use_rot: False

test_data:
  target: main.DataModuleFromConfig
  params:
    batch_size: 1
    num_workers: 6
    wrap: false
    test:
      target: basicsr.data.realesrgan_dataset.RealESRGANDataset
      params:
        gt_path: /mnt/lustre/share/jywang/dataset/ImageSR/DIV2K/DIV2K_train_HR/
        crop_size: 512
        io_backend:
          type: disk

        blur_kernel_size: 21
        kernel_list: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob: 0.1
        blur_sigma: [0.2, 1.5]
        betag_range: [0.5, 2.0]
        betap_range: [1, 1.5]

        blur_kernel_size2: 11
        kernel_list2: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
        kernel_prob2: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
        sinc_prob2: 0.1
        blur_sigma2: [0.2, 1.0]
        betag_range2: [0.5, 2.0]
        betap_range2: [1, 1.5]

        final_sinc_prob: 0.8

        gt_size: 512
        use_hflip: True
        use_rot: False

lightning:
  modelcheckpoint:
    params:
      every_n_train_steps: 1500
  callbacks:
    image_logger:
      target: main.ImageLogger
      params:
        batch_frequency: 1500
        max_images: 4
        increase_log_steps: False

  trainer:
    benchmark: True
    max_steps: 800000
    accumulate_grad_batches: 4
```

I wonder if this noisy loss curve is expected and where might I do wrong.

Thank you for your time and answer

**Comment by IceClear** [comment]
The loss is based on different time steps, so it is expected.
